[AOTI] fixed performance issue for AOTI_TORCH_CHECK

chenyang78 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128402

We introduced AOTI_TORCH_CHECK in #119220 to resolve slow-compilation
time issues. Unfortunately, it caused perf regressions for CPU
, as described in issue #126665. After some investigation, it turned
out the slow compilation was caused by the use of the builtin
function __builtin_expect provided by gcc/clang. Moreover,
nuking __builtin_expect doesn't seem to cause any performance penalty,
even though its purpose is to improve performance by providing the
compiler with branch prediction information.

abs latency numbers using the script shared by #126665:

                            before the fix      after the fix
T5Small                     1019.055694         917.875027
T5ForConditionalGeneration  1009.825196         916.369239

Commit:
[AOTI] fixed performance issue for AOTI_TORCH_CHECK

We introduced AOTI_TORCH_CHECK in #119220 to resolve slow-compilation
time issues. Unfortunately, it caused perf regressions for CPU
, as described in issue #126665. After some investigation, it turned
out the slow compilation was caused by the use of the builtin
function __builtin_expect provided by gcc/clang. Moreover,
nuking __builtin_expect doesn't seem to cause any performance penalty,
even though its purpose is to improve performance by providing the
compiler with branch prediction information.

abs latency numbers using the script shared by #126665:

                            before the fix      after the fix
T5Small                     1019.055694         917.875027
T5ForConditionalGeneration  1009.825196         916.369239

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -558,29 +558,24 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_proxy_executor_call_function(
-      TORCH_CHECK_MSG(cond, "", __VA_ARGS__));
+#define AOTI_TORCH_CHECK(cond, ...)              \
+  if (!(cond)) {                                 \
+    ::c10::detail::torchCheckFail(               \
```
I don't think we should directly refer to torchCheckFail here?

Taking a close look at https://github.com/pytorch/pytorch/blob/5d8c7f39d46699d8f8e92512309ea3499a29c08a/c10/util/Exception.cpp#L109, I think even the current way `aoti_torch_check` is written is problematic, because we don't want to throw an exception across the boundary here.  We should return `AOTITorchError` and check the error code at the call site, just like other C shim functions. For the purpose of this PR, I am ok with keeping `aoti_torch_check` as is but removing `C10_UNLIKELY_OR_CONST` in its implemenation. We can come up with a follow-up PR to fix the exception thrown issue.

chenyang78 replied:
Hmm, I think throwing an exception would be fine as we only use AOTI_TORCH_CHECK in the ```AOTIModel::run_impl```, which is called by ```AOTInductorModelContainerRun```. We convert all exceptions into error code when we cross the C interface. 

Since we are removing ```aoti_torch_check```, I am thinking perhaps we could move ```AOTI_TORCH_CHECK``` into utils.h so that it wouldn't be exposed via c_shim interface at all. What do you think?

chenyang78 replied:
Good point on the BC-breakage! I updated the diff to keep the ```aoti_torch_check``` and also put a comment to indicate we are going to deprecate it.

Regarding using ```std::runtime_error```, I think we could probably stick with the current use of ```::c10::detail::torchCheckFail```, which gives a nice error string upon any failure.

desertfire replied:
The point is `::c10::detail::torchCheckFail` should not appear in `c/shim.h`, unless `c10/util/Exception.h` is implemented in a header-only way. Let's chat offline.

desertfire commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -893,17 +893,6 @@ AOTITorchError aoti_torch_proxy_executor_call_function(
-    const char* file,
-    uint32_t line,
-    const char* msg) {
-  if (C10_UNLIKELY_OR_CONST(!cond)) {
```
Is `C10_UNLIKELY_OR_CONST` causing CPU benchmark perf regression? We could simply remove it here.

chenyang78 replied:
No. CPU benchmark perf regression is caused by *not* inlining the aoti_torch_check. ```C10_UNLIKELY_OR_CONST``` is the cause for the slow compilation that #119220 tried to address. So, simply removing ```C10_UNLIKELY_OR_CONST``` can fix the slow compilation issue as well. Moreover, removing ```C10_UNLIKELY_OR_CONST``` allows us to inline the error check code and hence fixes the CPU perf regression, too.

desertfire replied:
Got it.

Commit:
Update on "[AOTI] fixed performance issue for AOTI_TORCH_CHECK"


We introduced AOTI_TORCH_CHECK in #119220 to resolve slow-compilation
time issues. Unfortunately, it caused perf regressions for CPU
, as described in issue #126665. After some investigation, it turned
out the slow compilation was caused by the use of the builtin
function __builtin_expect provided by gcc/clang. Moreover,
nuking __builtin_expect doesn't seem to cause any performance penalty,
even though its purpose is to improve performance by providing the
compiler with branch prediction information.

abs latency numbers using the script shared by #126665:

                            before the fix      after the fix
T5Small                     1019.055694         917.875027
T5ForConditionalGeneration  1009.825196         916.369239

[ghstack-poisoned]

Commit:
Update on "[AOTI] fixed performance issue for AOTI_TORCH_CHECK"


We introduced AOTI_TORCH_CHECK in #119220 to resolve slow-compilation
time issues. Unfortunately, it caused perf regressions for CPU
, as described in issue #126665. After some investigation, it turned
out the slow compilation was caused by the use of the builtin
function __builtin_expect provided by gcc/clang. Moreover,
nuking __builtin_expect doesn't seem to cause any performance penalty,
even though its purpose is to improve performance by providing the
compiler with branch prediction information.

abs latency numbers using the script shared by #126665:

                            before the fix      after the fix
T5Small                     1019.055694         917.875027
T5ForConditionalGeneration  1009.825196         916.369239

[ghstack-poisoned]

Commit:
Update on "[AOTI] fixed performance issue for AOTI_TORCH_CHECK"


We introduced AOTI_TORCH_CHECK in #119220 to resolve slow-compilation
time issues. Unfortunately, it caused perf regressions for CPU
, as described in issue #126665. After some investigation, it turned
out the slow compilation was caused by the use of the builtin
function __builtin_expect provided by gcc/clang. Moreover,
nuking __builtin_expect doesn't seem to cause any performance penalty,
even though its purpose is to improve performance by providing the
compiler with branch prediction information.

abs latency numbers using the script shared by #126665:

                            before the fix      after the fix
T5Small                     1019.055694         917.875027
T5ForConditionalGeneration  1009.825196         916.369239

[ghstack-poisoned]

desertfire reviewed (approved):
cc @chunyuan-w

chenyang78 commented:
@pytorchbot merge
